// @prisma/codemods/bin/cli.js
"use strict";
/**
 * Copyright 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
// Based on https://github.com/reactjs/react-codemod/blob/dd8671c9a470a2c342b221ec903c574cf31e9f57/bin/cli.js
// @next/codemod optional-name-of-transform optional/path/to/src [...options]
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const globby_1 = __importDefault(require("globby"));
const inquirer_1 = __importDefault(require("inquirer"));
// @ts-ignore
const is_git_clean_1 = __importDefault(require("is-git-clean"));
const meow_1 = __importDefault(require("meow"));
const getCustomImportPath_1 = require("../utils/getCustomImportPath");
const runner_1 = require("../utils/runner");
function checkGitStatus(force) {
    let clean = false;
    let errorMessage = "Unable to determine if git directory is clean";
    try {
        clean = is_git_clean_1.default.sync(process.cwd());
        errorMessage = "Git directory is not clean";
    }
    catch (err) {
        if (err && err.stderr && err.stderr.indexOf("Not a git repository") >= 0) {
            clean = true;
        }
    }
    if (!clean) {
        if (force) {
            console.log(`WARNING: ${errorMessage}. Forcibly continuing.`);
        }
        else {
            console.log("Thank you for using @prisma/codemods!");
            console.log(chalk_1.default.yellow("\nBut before we continue, please stash or commit your git changes."));
            console.log("\nYou may use the --force flag to override this safety check.");
            process.exit(1);
        }
    }
}
const TRANSFORMER_INQUIRER_CHOICES = [
    {
        name: "update-2.12: Runs namespace/findUnique/to$ transforms",
        value: "update-2.12",
    },
    {
        name: "namespace: Codemod for @prisma/client namespace change",
        value: "namespace",
    },
    {
        name: "to$: Converts deprecated prisma.x methods to prisma.$x",
        value: "to$",
    },
    {
        name: "findUnique: Codemod for @prisma/client that converts findOne to findUnique",
        value: "findUnique",
    },
];
function expandFilePathsIfNeeded(filesBeforeExpansion) {
    const shouldExpandFiles = filesBeforeExpansion.some((file) => file.includes("*"));
    return shouldExpandFiles
        ? globby_1.default.sync(filesBeforeExpansion)
        : filesBeforeExpansion;
}
function run() {
    const cli = meow_1.default(`
  Usage
    $ npx @prisma/codemods <transform> <path> <...options>
      transform    One of the choices from https://github.com/prisma/codemods#transforms
      path         Directory of your app. i.e ./my-awesome-project
  Options
    --force            Bypass Git safety checks and forcibly run codemods
    --schemaPath       Specify a path to your ./prisma/schema.prisma
    --dry              Dry run (no changes are made to files)
    --print            Print transformed files to your terminal
`, {
        autoHelp: true,
        flags: {
            schemaPath: {
                type: "string",
                isRequired: false,
            },
            force: {
                alias: "f",
                type: "boolean",
            },
            dry: {
                alias: "d",
                type: "boolean",
            },
            print: {
                alias: "p",
                type: "boolean",
            },
            runInBand: {
                type: "boolean",
            },
            help: {
                alias: "h",
                type: "boolean",
            },
        },
    });
    if (!cli.flags.dry) {
        checkGitStatus(cli.flags.force);
    }
    if (cli.input[0] &&
        !TRANSFORMER_INQUIRER_CHOICES.find((x) => x.value === cli.input[0])) {
        console.error("Invalid transform choice, pick one of:");
        console.error(TRANSFORMER_INQUIRER_CHOICES.map((x) => "- " + x.value).join("\n"));
        process.exit(1);
    }
    inquirer_1.default
        .prompt([
        {
            type: "input",
            name: "dir",
            message: "On which directory should the codemods be applied?",
            when: !cli.input[1],
            default: ".",
            // validate: () =>
            filter: (files) => files.trim(),
        },
        {
            type: "list",
            name: "transformer",
            message: "Which transform would you like to apply?",
            when: !cli.input[0],
            pageSize: TRANSFORMER_INQUIRER_CHOICES.length,
            choices: TRANSFORMER_INQUIRER_CHOICES,
        },
    ])
        .then((answers) => __awaiter(this, void 0, void 0, function* () {
        const { dir, transformer } = answers;
        const dirPath = cli.input[1] || dir;
        const selectedTransformer = cli.input[0] || transformer;
        process.env.PRISMA_CUSTOM_IMPORT_PATH = yield getCustomImportPath_1.getCustomImportPath({
            schemaPathFromArgs: cli.flags.schemaPath,
            cwd: dirPath,
        });
        if (selectedTransformer === 'update-2.12') {
            const namespace = yield runner_1.runTransform({
                projectDir: dirPath,
                flags: cli.flags,
                customImportPath: process.env.PRISMA_CUSTOM_IMPORT_PATH,
                transformer: 'namespace',
            });
            const findUnique = yield runner_1.runTransform({
                projectDir: dirPath,
                flags: cli.flags,
                customImportPath: process.env.PRISMA_CUSTOM_IMPORT_PATH,
                transformer: 'findUnique',
            });
            const to$ = yield runner_1.runTransform({
                projectDir: dirPath,
                flags: cli.flags,
                customImportPath: process.env.PRISMA_CUSTOM_IMPORT_PATH,
                transformer: 'to$',
            });
        }
        else {
            return runner_1.runTransform({
                projectDir: dirPath,
                flags: cli.flags,
                customImportPath: process.env.PRISMA_CUSTOM_IMPORT_PATH,
                transformer: selectedTransformer,
            });
        }
    }));
}
module.exports = {
    run: run,
    runTransform: runner_1.runTransform,
    checkGitStatus: checkGitStatus,
    jscodeshiftExecutable: runner_1.jscodeshiftExecutable,
    transformerDirectory: runner_1.transformerDirectory,
};
